accepted/exception-handling.md

```diff
@@ -0,0 +1,240 @@
+  into a `catch` block. Instead, the control flow edges to `catch`
+  must come via a `try_call` instruction.
+
+* A new call instruction `try_call <ok_label>, <exception_label>`,
+  reminiscent of LLVM's `invoke`, where `<ok_label>` must be a regular
+  block and `<exception_label>` must be `catch` block. The block
+  parameters of `<ok_label>` must match the return types of the called
+  function. The semantics is as follows: when `try_call` returns
+  normally, control tranferred to the block named by `<ok_label>`;
+  when `try_call` is unwound, control is tranferred to the block named
+  by `<exception_label>` (note this may happen multiple times in a
+  two-phase exceptions scenario).
+
+* Potentially we may also want a `try_call_indirect <ok_label>,
+  <exception_label>` for indirect calls.
+
+We do not define a CLIF instruction for throwing an
+exception. Instead, exception throwing must be done indirectly via an
+imported function (e.g. a Wasmtime builtin libcall implemented in the
+host/engine).
+
+We do not hard-wire in support for two-phase exceptions. Though, it
+should be possible to encode two-phase exceptions on top of the
+proposed constructs. For example, the producer can emit some prelude
+code that runs in the beginning of each `catch` to determine whether
+the runtime is in the search phase or unwind phase.
+
+## Implementation strategies
+
+There are at least three feasible implementation strategies for stack
+unwinding.
+
+1. Side table: we can store information about which parts of the code
+can raise exceptions and where their respective handlers are located
+in a bespoke runtime-managed data structure.
+
+2. Calling convention: we can extend the calling convention to
+propagate exception information between function calls.
+
+3. DWARF unwinder: we can implement a bespoke unwinder for a subset of
+DWARF. This subset should cover the set of special-purpose registers
+used by Wasmtime.
+
+### Incremental strategy for zero-cost exceptions
+
+We are interested in landing this feature in a reasonably timely
+manner to enable producers to use exception handling and allow
+dependent proposal to run experiments on top of Wasmtime. Therefore we
+propose to start by implementing exceptions using calling convention
+based strategy. The strategy we have in mind is not zero-cost, but
+possibly cheap enough that is acceptable in the interim.
+
+Concretely, we propose to implement non-zero cost exceptions by using
+flags (e.g. the overflow flag) to determine whether a call returned
+through a throw or an ordinary return instruction. Put into code:
+
+```
+call 0x12345678 # exceptional return sets the flag
+jo $exception_branch
+```
+
+Before returning normally any function must clear the flag, e.g.
+
+```
+test al, al
+ret
+```
+
+We reckon this approach is relatively low overhead, and it something
+we can confidently implement correctly more quickly than the side
+table or DWARF unwinder strategies. Adopting this strategy would allow
```

Comment:
One thing worth pointing out about this approach is that given the non-zero-cost in the non-exceptional case it will likely prevent turning this proposal on by default. The cost would be incurred for users who don't use exceptions at all unless Wasmtime implements a form of detection of exception-using-instructions which I think could get particularly hairy in the cross-instance semantics below.